gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/GCImpl.java

```diff
@@ -0,0 +1,125 @@
+  private final Map<String, Instant> cutOffTimestampPerRef;
+
+  public GCImpl(
+      Map<String, String> configuration,
+      Instant defaultCutOffTimeStamp,
+      Map<String, Instant> cutOffTimestampPerRef) {
+    Objects.requireNonNull(configuration);
+    Objects.requireNonNull(defaultCutOffTimeStamp);
+    this.configuration = configuration;
+    this.defaultCutOffTimeStamp = defaultCutOffTimeStamp;
+    this.cutOffTimestampPerRef =
+        cutOffTimestampPerRef == null ? new HashMap<>() : cutOffTimestampPerRef;
+  }
+
+  public IdentifiedResult identifyExpiredContents(SparkSession session) {
+    // a) Algorithm for identifying the live contents and return the bloom filter per content-id.
+    //    Walk through each live reference distributively (one spark task for each reference).
+    //    For each commit in a reference, collect dropped table keys if present.
+    //    If the commit is head commit, get all keys and add its contents to bloom filter.
+    //    Else If for the keys that are NOT in dropped table keys AND commit is not expired (based
+    //      on cutoff time of this reference). Add its content to bloom filter.
+    //    Else if the content is expired. Stop traversing as there cannot be any more live commits
+    //      for this reference.
+    //    Collect bloom filter per content id from each task and merge them.
+    NessieApiV1 api = DistributedIdentifyUtil.getApi(configuration);
+    List<Reference> references = api.getAllReferences().get().getReferences();
+    DistributedIdentifyUtil distributedIdentifyUtil =
+        new DistributedIdentifyUtil(session, configuration);
+    Map<String, ContentBloomFilter> liveContentsBloomFilterMap =
+        distributedIdentifyUtil.getLiveContents(
+            references, cutOffTimestampPerRef, defaultCutOffTimeStamp);
+
+    // b) Algorithm for identifying the expired and live contents
+    //      and return the list of expired contents and list of live contents per content id.
+    //    Get the list of dead reference from reflog
+    //      and append it to current list of live references.
+    //    Walk through each reference distributively (one spark task for each reference).
+    //    For each commit in a reference,
+    //      get the contents (use DETACHED reference here as some references are dead).
+    //    Check if the content is expired by checking against the bloom filter.
+    //    Merge the result of each task and return the result.
+    List<Reference> deadReferences = collectDeadReferences(api);
+    List<Reference> allRefs = new ArrayList<>(references);
+    if (deadReferences.size() > 0) {
+      allRefs.addAll(deadReferences);
+    }
+    return distributedIdentifyUtil.getIdentifiedResults(liveContentsBloomFilterMap, allRefs);
+  }
+
+  private List<Reference> collectDeadReferences(NessieApiV1 api) {
+    List<RefLogResponse.RefLogResponseEntry> reflogResponseEntries;
+    try {
+      reflogResponseEntries =
```

Comment:
Any suggestions ? I couldn't find a better way at the moment.

Reply:
we might have to periodically clean up reflog too then.
Let's see what @rymurr thinks.

I even have a crazy idea of adding a marker row in reflog for every 250 entries. So that I can query the marker rows first using database filter query and then distributively reflog traversal per between marker row reflog ids for faster results. But may be too much for now.